presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java

```diff
@@ -239,6 +251,44 @@ private <C> void createQueryInternal(QueryId queryId, String slug, int retryCoun
         }
     }
 
+    private Session checkPermissions(Session session)
+    {
+        // When selectAuthorizedIdentity API is not enabled, only check delegation permission
+        if (!securityConfig.isAuthorizedIdentitySelectionEnabled()) {
+            accessControl.checkCanSetUser(
+                    session.getIdentity(),
+                    new AccessControlContext(
+                            session.getQueryId(),
+                            session.getClientInfo(),
+                            session.getSource()),
+                    session.getIdentity().getPrincipal(),
+                    session.getIdentity().getUser());
+            return session;
+        }
+
+        // When selectAuthorizedIdentity API is enabled,
+        // 1. Check the delegation permission, which is inside the API call
+        // 2. Select the authorized user
+        AuthorizedIdentity authorizedIdentity = accessControl.selectAuthorizedIdentity(
+                session.getIdentity(),
+                new AccessControlContext(
+                        session.getQueryId(),
+                        session.getClientInfo(),
+                        session.getSource()),
+                session.getIdentity().getUser(),
+                session.getCertificates());
+
+        // 3. Update the identity accordingly in the session
+        Identity updatedIdentity = new Identity(
```

Comment:
Since our first version will never update the identity, do we need to create new identity object?

Reply:
Ok, that is fine then. We need update it whenever we want. Will remove this part.